impl/spanner/mutations/table.go

```diff
@@ -0,0 +1,311 @@
+	// TODO: Replace with MAX(Revision) when spansql supports aggregate operators.
+	stmt := spanner.NewStatement(`
+	SELECT Timestamp FROM Mutations WHERE
+		DirectoryID = @directoryID AND
+		LogID = @logID AND
+		Timestamp >= @start
+	ORDER BY Timestamp
+	LIMIT @limit`)
+	stmt.Params["directoryID"] = directoryID
+	stmt.Params["logID"] = logID
+	stmt.Params["start"] = markToTime(start)
+	stmt.Params["limit"] = int64(batchSize)
+
+	var cnt int64
+	var max time.Time
+	rtx := t.client.Single().WithTimestampBound(spanner.MaxStaleness(readStaleness))
+	defer rtx.Close()
+	if err := rtx.Query(ctx, stmt).Do(
+		func(row *spanner.Row) error {
+			cnt++
+			var watermark time.Time
+			if err := row.Column(0, &watermark); err != nil {
+				return err
+			}
+			if watermark.After(max) {
+				max = watermark
+			}
+			return nil
+		}); err != nil {
+		return 0, water.Mark{}, err
+	}
+	if cnt == 0 {
+		// When there are no rows, return the start time as the highest timestamp.
+		return 0, start, nil
+	}
+	return int32(cnt), timeToMark(max).Add(1), nil
+}
+
+// ReadLog reads all mutations between [low, high).
+func (t *Table) ReadLog(ctx context.Context, directoryID string, logID int64, low, high water.Mark,
+	limit int32) ([]*mutator.LogMessage, error) {
+	if high.Value() == 0 || limit == 0 {
+		return []*mutator.LogMessage{}, nil
+	}
+
+	msgs := make([]*mutator.LogMessage, 0)
+	rtx := t.client.ReadOnlyTransaction().
+		WithTimestampBound(spanner.MinReadTimestamp(markToTime(high)))
+	defer rtx.Close()
+	if err := rtx.ReadWithOptions(ctx, mutTable,
+		spanner.KeyRange{
+			Kind:  spanner.ClosedOpen,
+			Start: spanner.Key{directoryID, logID, markToTime(low), 0},
+			End:   spanner.Key{directoryID, logID, markToTime(high), 0},
+		},
+		[]string{"Timestamp", "LocalID", "Mutation"},
+		&spanner.ReadOptions{Limit: int(limit)},
+	).Do(func(r *spanner.Row) error {
+		msg, err := unmarshalRow(r, logID)
+		if err != nil {
+			return err
+		}
+		msgs = append(msgs, msg)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	var maxTs water.Mark
+	var maxLocal int64
+	for _, m := range msgs {
+		if m.ID.Value() > maxTs.Value() {
+			maxTs = m.ID
+			maxLocal = 0
+		}
+		if m.ID.Value() == maxTs.Value() &&
+			m.LocalID > maxLocal {
+			maxLocal = m.LocalID
+		}
+	}
+	if len(msgs) == int(limit) {
+		// There might be more messages in this batch to read.
+		if err := rtx.Read(ctx, mutTable,
+			spanner.KeyRange{
+				Kind:  spanner.OpenOpen,
+				Start: spanner.Key{directoryID, logID, markToTime(maxTs), maxLocal},
+				End:   spanner.Key{directoryID, logID, markToTime(maxTs.Add(1)), 0},
+			},
+			[]string{"Timestamp", "LocalID", "Mutation"},
+		).Do(func(r *spanner.Row) error {
+			msg, err := unmarshalRow(r, logID)
+			if err != nil {
+				return err
+			}
+			msgs = append(msgs, msg)
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
This whole block is surprising to me. I would expect `limit` to be the max number to return, but it appears to not be that. I think some doc string on what it means would be helpful. Or this is an indication that I need to eat now :-)

Reply:
This certainly deserves documentation. 

This is an edge case I'm not super happy with, but fixing it requires a bit of an overhaul -- and would reduce our ability to control traffic to the map.

Reply:
I understand this now and think it's reasonable. I think the missing concept for me was that of the "batch". It almost seems like Send should be called WriteBatch.